src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs

```diff
@@ -158,6 +158,23 @@ internal static partial class SyntaxGeneratorExtensions
                     return generatorInternal.IsPatternExpression(leftOperand, operatorToken, generatorInternal.ConstantPattern(generator.NullLiteralExpression()));
                 }
 
+                // With a not pattern, it is possible to have an ambiguity between a type name or a local identifier.
+                // For example, if a class is named C and we have:
+                // C C = new();
+                // object O = C;
+                // if (O is not C)
+                // ...
+                // Then there is an ambiguity between the type C and the local named C.
+                // To address this, we use an expression with the fully qualified type name which the simplifier
+                // will shorten to the shortest possible without ambiguity
+                if (!syntaxFacts.IsPredefinedType(rightOperand) &&
```

Comment:
i think a better fix is to get the typeNode back, and then determine what sort of pattern needs to be generated based on that typeNode.

Reply:
Sorry, I am confused on how one would do it here: the original expression at best gives me a PredefinedTypeSyntax or an IdentifierNameSyntax. That is the only information I have here to determine which pattern it should be and IsPredefinedType covers this check.

If it's not a predefined type, I have to generate a fully qualified name of the type and wrap it in a constant pattern. The typeNode is only involved there because this is when I know the format it should have.

I think I am missunderstanding because it sounds like you're asking to obtain something I can't get or know if I should get before checking what type rightOperand is.

Reply:
can i make edits to your PR?

Reply:
Sure, go ahead.